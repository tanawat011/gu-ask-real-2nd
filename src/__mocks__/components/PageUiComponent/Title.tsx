// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck

export const Title = () => {
  return <div data-testid='title'>Title</div>
}
