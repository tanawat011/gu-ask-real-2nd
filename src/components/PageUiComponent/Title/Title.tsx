import tw, { styled } from 'twin.macro'

type TitleProps = {
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | '2xl'
  title: string
  description?: string | React.ReactNode
}

type TwTitleSize = Pick<TitleProps, 'size'>

const TwContainer = styled.div(({ size }: TwTitleSize) => {
  switch (size) {
    case '2xl':
      return [tw`mb-12`]
    case 'xl':
      return [tw`mb-10`]
    case 'lg':
      return [tw`mb-8`]
    case 'md':
      return [tw`mb-6`]
    case 'sm':
      return [tw`mb-4`]
    case 'xs':
      return [tw`mb-3`]
  }
})

export const Title: React.FC<TitleProps> = ({ description, title, size = 'md' }) => {
  const renderTitleSize = () => {
    switch (size) {
      case '2xl':
        return <h1>{title}</h1>
      case 'xl':
        return <h2>{title}</h2>
      case 'lg':
        return <h3>{title}</h3>
      case 'md':
        return <h4>{title}</h4>
      case 'sm':
        return <h5>{title}</h5>
      case 'xs':
        return <h6>{title}</h6>
    }
  }

  return (
    <TwContainer size={size}>
      {renderTitleSize()}
      <span>{description}</span>
    </TwContainer>
  )
}
