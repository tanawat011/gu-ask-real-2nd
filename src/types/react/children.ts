export type Children = { children: JSX.Element }
