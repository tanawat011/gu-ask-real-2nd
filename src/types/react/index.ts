export * from './children'
export * from './reactRouterDom'
