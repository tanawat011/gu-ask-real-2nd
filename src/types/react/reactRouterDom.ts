import type { Scroll } from 'types/common'

export type OutletContextProps = {
  scroll: Scroll
}
