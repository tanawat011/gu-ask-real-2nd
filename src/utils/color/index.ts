export * from './hex2Rgb'
export * from './rgb2Hex'
