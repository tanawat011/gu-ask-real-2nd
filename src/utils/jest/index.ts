export * from './renderWithProviders'
export * from './twColor'
