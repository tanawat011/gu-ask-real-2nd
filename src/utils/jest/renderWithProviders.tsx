import type { RenderResult } from '@testing-library/react'

import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { RecoilRoot } from 'recoil'

const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  const queryClient = new QueryClient()

  return (
    <QueryClientProvider client={queryClient}>
      <RecoilRoot>
        <BrowserRouter>{children}</BrowserRouter>
      </RecoilRoot>
    </QueryClientProvider>
  )
}

interface RenderWithProvidersOptions {
  router?: {
    route: string
  }
  recoil?: object
}

export const renderWithProviders = (
  ui: JSX.Element,
  option?: RenderWithProvidersOptions,
): RenderResult => {
  const routerOption = { ...(option?.router || { route: '/' }) }
  window.history.pushState({}, 'Test page', routerOption.route)

  return {
    ...render(ui, { wrapper: AllTheProviders, ...routerOption }),
  }
}
