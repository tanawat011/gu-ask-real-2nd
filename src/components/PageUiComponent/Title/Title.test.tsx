import { screen } from '@testing-library/react'

import { renderWithProviders } from 'utils/jest'

import { Title } from './Title'

jest.mock('./Title', () => jest.requireActual('./Title'))

describe('<Title />', () => {
  test('renders correctly with default size', () => {
    expect.assertions(4)

    const { asFragment } = renderWithProviders(<Title title='Title' description='Detail' />)

    const firstElement = screen.getByTestId('container')
    const headTitle = screen.getByRole('heading', { level: 4 })

    expect(firstElement).toHaveStyleRule('margin-bottom', '1.5rem')
    expect(headTitle).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders correctly with size `xs`', () => {
    expect.assertions(3)

    renderWithProviders(<Title title='Title' size='xs' />)

    const firstElement = screen.getByTestId('container')
    const headTitle = screen.getByRole('heading', { level: 6 })

    expect(firstElement).toHaveStyleRule('margin-bottom', '0.75rem')
    expect(headTitle).toBeInTheDocument()
  })

  test('renders correctly with size `sm`', () => {
    expect.assertions(3)

    renderWithProviders(<Title title='Title' size='sm' />)

    const firstElement = screen.getByTestId('container')
    const headTitle = screen.getByRole('heading', { level: 5 })

    expect(firstElement).toHaveStyleRule('margin-bottom', '1rem')
    expect(headTitle).toBeInTheDocument()
  })

  test('renders correctly with size `md`', () => {
    expect.assertions(3)

    renderWithProviders(<Title title='Title' size='md' />)

    const firstElement = screen.getByTestId('container')
    const headTitle = screen.getByRole('heading', { level: 4 })

    expect(firstElement).toHaveStyleRule('margin-bottom', '1.5rem')
    expect(headTitle).toBeInTheDocument()
  })

  test('renders correctly with size `lg`', () => {
    expect.assertions(3)

    renderWithProviders(<Title title='Title' size='lg' />)

    const firstElement = screen.getByTestId('container')
    const headTitle = screen.getByRole('heading', { level: 3 })

    expect(firstElement).toHaveStyleRule('margin-bottom', '2rem')
    expect(headTitle).toBeInTheDocument()
  })

  test('renders correctly with size `xl`', () => {
    expect.assertions(3)

    renderWithProviders(<Title title='Title' size='xl' />)

    const firstElement = screen.getByTestId('container')
    const headTitle = screen.getByRole('heading', { level: 2 })

    expect(firstElement).toHaveStyleRule('margin-bottom', '2.5rem')
    expect(headTitle).toBeInTheDocument()
  })

  test('renders correctly with size `2xl`', () => {
    expect.assertions(3)

    renderWithProviders(<Title title='Title' size='2xl' />)

    const firstElement = screen.getByTestId('container')
    const headTitle = screen.getByRole('heading', { level: 1 })

    expect(firstElement).toHaveStyleRule('margin-bottom', '3rem')
    expect(headTitle).toBeInTheDocument()
  })
})
