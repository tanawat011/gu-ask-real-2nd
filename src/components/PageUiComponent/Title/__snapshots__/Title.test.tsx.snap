// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Title /> renders correctly with default size 1`] = `
<DocumentFragment>
  .c0 {
  margin-bottom: 1.5rem;
}

<div
    class="c0"
    data-testid="container"
  >
    <h4>
      Title
    </h4>
    <span>
      Detail
    </span>
  </div>
</DocumentFragment>
`;
